### PR TITLE
Log warning when using central packagement management but not package source mapping

### LINKED ISSUE #11505: [Feature]: Require package source mapping when using CPM
### NuGet Product(s) Involved

NuGet.exe, Visual Studio Package Management UI, Visual Studio Package Manager Console, MSBuild.exe, dotnet.exe

### The Elevator Pitch

To make NuGet secure by default, there should be an option to required package source mapping when using more than one feed and central package management.

### Additional Context and Details

_No response_

### PR DESCRIPTION
**Pull request by jeffkl**
## Bug

Fixes: https://github.com/NuGet/Home/issues/11505

Regression? Last working version:

## Description
Log a warning during restore if central package management is being used, there is more than one feed configured, and package source mapping is not enabled.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [x] Documentation PR or issue filled
  - **OR**
  - [ ] N/A

### CODE REVIEW

File: src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs]: Same log NU1504 used in https://github.com/NuGet/NuGet.Client/pull/4484

Comment (jeffkl) [reply src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs]: I swear I looked at #4484 and it wasn't using NU1504...  I'll update my code

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs
Comment (erdembayar) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs]: Is having more than 1 source is bad practice with central package management?
Still trying to understand why it's the issue.

File: src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs]: Most likely he changed recently.
image

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs]: Using just this config probably breaks the redirection of the global packages folder, and likely makes the project "leak" packages into the user's global packages folder. 

Probably best to edit the existing config.  
https://github.com/NuGet/NuGet.Client/blob/dev/test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestSettingsContext.cs

You can add helper if there are none.

**Review by nkolev92 — COMMENTED** [review]
Change looks great, the test probably need cleaned up.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:474]: I think we should raise the warning only if more than 1 http source(s) are configured. With the current logic, warning is raised if there are 2 sources (1 http source + 1 local feed) configured. Please let me know if I am missing something.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs
Comment (jeffkl) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs]: We want to make NuGet secure by default and in this case since the customer is using CPM, they should at least be told they can use package source mapping to be more secure.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:474]: Good point, I would assume that multiple feeds can always lead to restore non-determinism.  But I could see the argument of only checking HTTP feeds.  What does everyone else think?

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs
Comment (jeffkl) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs]: Good call, I'll use the `internal` constructor instead

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:474]: The only scenario in which you're likely to have a deterministic behavior by accident is a 2 source configuration where you have 1 local and 1 remote feed. The local feeds would get first dibs on every id being acquired, only if exact version is required. 

I don't think we should add exceptions. as Jeff mentioned you're not always guaranteed determinism any time you have more than 1 feed.

Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:474]: Thanks for the details. Even though source mapping is enabled, the user can configure the package patterns in such a way that one or more sources may be queried for a particular package. In this case, we may also have to raise a warning during restore saying that `multiple sources are configured in source mapping configuration for {package id} ` when CPM is enabled. May be in a follow-up PR. What does everyone else think?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:474]: > Even though source mapping is enabled, the user can configure the package patterns in such a way that one or more sources may be queried for a particular package. In this case, we may also have to raise a warning during restore saying that multiple sources are configured in source mapping configuration for {package id} when CPM is enabled.

Personally, I'd argue this is a separate, source mapping only concern.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:474]: In fact, it's basically what used to be strict mode, and our direction for that right now is to wait for feedback.

File: src/NuGet.Core/NuGet.Commands/Strings.resx
Comment (JonDouglas) [inline src/NuGet.Core/NuGet.Commands/Strings.resx]: There are {0} package sources defined in your configuration. To enable central package management, please map your package sources with {aka.ms link} or specify a single package source to continue.

{0} - Number of sources.

{aka.ms link} - https://docs.microsoft.com/en-us/nuget/consume-packages/package-source-mapping

@nkolev92 thoughts?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/Strings.resx]: Love it.

**Comment by jeffkl** [comment]
Team Review: Decide if we want to allow the feeds that Visual Studio configures so that people don't get the warning from feeds configured not by them

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:474]: I looked into it more and as long as you have a `NuGet.Config` with a `<clear />` and single `<add />` then you get a single source no matter what other configs try to inject feeds.  So I think checking for more than one feed is good.  If we get feedback later on to ignore non-HTTP feeds we can incorporate that later.  Since this a new warning, people can of course suppress it and move on.

File: src/NuGet.Core/NuGet.Commands/Strings.resx
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/Strings.resx]: @nkolev92  do we already have an aka.ms link to package source mapping or should I just embed the URL in the message?

Comment (JonDouglas) [reply src/NuGet.Core/NuGet.Commands/Strings.resx]: I can create a link as I synced with nikolche and we use aka.ms for a number of NuGet and sdk things. What alias do we want @nkolev92?

nugetsourcemapping?
packagesourcemapping?
etc

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/Strings.resx]: either one works. Is there a pattern we have with some of our other aka.ms links?

Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/Strings.resx]: Gentle ping, I'd like to merge this today.  I can merge it for now with the link directly to the docs or we need to decide on the aka.ms link and I can use that even if it doesn't exist yet

Comment (JonDouglas) [reply src/NuGet.Core/NuGet.Commands/Strings.resx]: https://aka.ms/nuget-package-source-mapping

Comment (JonDouglas) [reply src/NuGet.Core/NuGet.Commands/Strings.resx]: Or if you want to use the docs link, that's fine too. Just remove the locale `en-us`. if you decide that way.

Comment (jeffkl) [reply src/NuGet.Core/NuGet.Commands/Strings.resx]: Just saw it come through, I've updated the PR to have https://aka.ms/nuget-package-source-mapping

**Comment by jeffkl** [comment]
@NuGet/nuget-client please review again with the final wording of the warning, @aortiz-msft does this look good to you?

**Review by nkolev92 — DISMISSED** [review]

**Comment by jeffkl** [comment]
@NuGet/nuget-client Had to rebase to get a green build which reset approvals, @nkolev92 please approve again

**Review by nkolev92 — APPROVED** [review]

**Review by JonDouglas — APPROVED** [review]
Ship it!

**Pull request merged**
